add world-position-changed to CCNode

knoxHuang opened the pull request:
Re: cocos-creator/fireball#5889

Changes proposed in this pull request:
 * 添加 ‘world-position-changed’ 事件

@cocos-creator/engine-admins

zilongshanren commented on cocos2d/core/CCNode.js:
```diff
@@ -442,6 +447,9 @@ var Node = cc.Class({
                                 this.emit(POSITION_CHANGED);
                             }
                         }
+
```
每次节点坐标变化都发这个事件？这样会卡爆吧？

jareguo reviewed (changes requested):
这个开销太大了，而且 setPosition, setPositionX 什么的 API 也没发送消息

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -1219,6 +1282,8 @@ var Node = cc.Class({
         this._checkTouchListeners();
         this._checkMouseListeners();
+
+        this._removeWorldPositionChanged();
```
targetOff 不需要这样吧，就算调用了这个方法，里面 hasListener 的判断都过不去。

knoxHuang replied:
过得去吧，在这个之前不是执行了 this._EventTargetTargetOff(target); ?

jareguo replied:
哦，我没看到~ 但这也还是不需要调用 _removeWorldPositionChanged 啊？

knoxHuang replied:
移除节点是不会走 off ，只会走 targetOff，如果不在这里执行，那你要如何 off 掉事件？

jareguo commented on cocos2d/core/utils/base-node.js:
```diff
@@ -384,6 +384,8 @@ var BaseNode = cc.Class({
             this._onHierarchyChanged(null);
         }
+
+        this.emit('parent-changed', oldParent);
```
用宏

jareguo commented on test/qunit/unit-es5/test-world-position.js:
```diff
@@ -0,0 +1,147 @@
+module('world-position-changed');
+
+test('world position changed for parnet', function () {
```
world-position-changed should be emitted by parent

jareguo commented on test/qunit/unit-es5/test-world-position.js:
```diff
@@ -0,0 +1,147 @@
+    CC.parent = BB;
+    DD.parent = CC;
+    EE.parent = DD;
+    FF.parent = EE;
```
不用嵌套这么深…… 三层就够了，三层可以没理由 5 层不行

jareguo commented on test/qunit/unit-es5/test-world-position.js:
```diff
@@ -0,0 +1,147 @@
+    var called = false;
+    var _onWorldPositionChangedCallback = function () {
+        called = true;
+    };
```
把这个闭包替换成
var _onWorldPositionChangedCallback = new Callback().enable();

jareguo commented on test/qunit/unit-es5/test-world-position.js:
```diff
@@ -0,0 +1,147 @@
+
+    AA.position = cc.p(100, 100);
+
+    ok(called, 'FF receives world-position-changed event for parent position changed');
```
然后这里改成 _onWorldPositionChangedCallback.once('world-position-changed emitted by ?which? parent when position changed');

jareguo commented on test/qunit/unit-es5/test-world-position.js:
```diff
@@ -0,0 +1,147 @@
+    FF.off('world-position-changed', _onWorldPositionChangedCallback);
+
+    called = false;
+    EE.on('world-position-changed', _onWorldPositionChangedCallback);
```
不要反复修改监听的子节点，而是在同一个子节点上监听，父亲、祖父、曾祖父上的各种事件

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -36,6 +36,10 @@ var ANCHOR_CHANGED = 'anchor-changed';
+
+var PARENT_CHANGED = 'parent-changed';
+
 var CHILD_ADDED = 'child-added';
```
WORLD_POSITION_CHANGED 放在 POSITION_CHANGED 后一行

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -36,6 +36,10 @@ var ANCHOR_CHANGED = 'anchor-changed';
+var WORLD_POSITION_CHANGED = 'world-position-changed';
+
+var PARENT_CHANGED = 'parent-changed';
+
```
PARENT_CHANGED 之后没必要空一行吧…… 和 CHILD_ADDED 是同一批事件

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -1091,6 +1095,58 @@ var Node = cc.Class({
+        // remove oldParent event
+        var oldParent = event.detail;
+        if (oldParent) {
+            oldParent.off(WORLD_POSITION_CHANGED, oldParent._onWorldPositionChanged, this);
```
传入的参数是 oldParent._onWorldPositionChanged 和 this，这逻辑说不通啊？
要获取的是 oldParent 示例上的成员方法，但是 this 又是 this…… 这是什么逻辑？

knoxHuang replied:
这里应该是写 oldParent.off(WORLD_POSITION_CHANGED, this._onWorldPositionChanged, this);

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -1091,6 +1095,58 @@ var Node = cc.Class({
+        if (oldParent) {
+            oldParent.off(WORLD_POSITION_CHANGED, oldParent._onWorldPositionChanged, this);
+            oldParent.off(PARENT_CHANGED, oldParent._onParentChanged, this);
+            oldParent._removeWorldPositionChanged();
```
oldParent.off(WORLD_POSITION_CHANGED) 时，本来就会触发 oldParent._removeWorldPositionChanged 了，这里不用重复调用

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -1095,55 +1093,43 @@ var Node = cc.Class({
-            oldParent.off(PARENT_CHANGED, oldParent._onParentChanged, this);
-            oldParent._removeWorldPositionChanged();
+            oldParent.off(WORLD_POSITION_CHANGED, this._onWorldPositionChanged, this);
+            oldParent.off(PARENT_CHANGED, this._registerWorldPosChangedToNewParent, this);
```
都没 on，怎么就 off 了

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -2576,6 +2560,10 @@ if (CC_JSB) {
  */
 /**
+ * @event world-position-changed
+ * @param {Event.EventCustom} event
```
又没有参数…… 把 event 去掉吧

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -2585,6 +2573,10 @@ if (CC_JSB) {
  */
 /**
+ * @event parent-changed
+ * @param {Event.EventCustom} event
```
有参数啊，oldParent

jareguo commented on cocos2d/core/CCNode.js:
```diff
@@ -1095,55 +1093,43 @@ var Node = cc.Class({
-            parent.off(SCALE_CHANGED, this._onWorldPositionChanged, this);
+            parent.onoff(register, WORLD_POSITION_CHANGED, this._onWorldPositionChanged, this);
+            parent.onoff(register, ROTATION_CHANGED, this._onWorldPositionChanged, this);
+            parent.onoff(register, SCALE_CHANGED, this._onWorldPositionChanged, this);
```
上面那三行代码出现了三次…… 提取一下吧？然后这里就 this._doRegisterWPCTo(parent, register);

knoxHuang committed:
add world-position-changed event to CCNode

add onoff to event-target
add parent-changed event  to CCNode
add test-world-position to test

jareguo reviewed (approved):
LGTM，本月新的最佳 PR 诞生了

zilongshanren commented on cocos2d/core/event/event-target.js:
```diff
@@ -152,6 +152,30 @@ proto.hasEventListener = function (type, checkCapture) {
+ *                              When false, callback will NOT be invoked when event's eventPhase attribute value is CAPTURING_PHASE.
+ *                              Either way, callback will be invoked when event's eventPhase attribute value is AT_TARGET.
+ */
+proto.onoff = function (isOn, type, callback, target, useCapture) {
```
这个 api 太扯了吧。。。 onoff 使用成本比单独使用 On或者 off 都要高。。。而且这个函数一下子就 5 个参数了。。

jareguo replied:
不然你有其它建议？比如 CCButton 中，
```
            this.node.off(cc.Node.EventType.TOUCH_START, this._onTouchBegan, this);
            this.node.off(cc.Node.EventType.TOUCH_MOVE, this._onTouchMove, this);
            this.node.off(cc.Node.EventType.TOUCH_END, this._onTouchEnded, this);
            this.node.off(cc.Node.EventType.TOUCH_CANCEL, this._onTouchCancel, this);

            this.node.off(cc.Node.EventType.MOUSE_ENTER, this._onMouseMoveIn, this);
            this.node.off(cc.Node.EventType.MOUSE_LEAVE, this._onMouseMoveOut, this);
```
完全和 _registerEvent 重复的代码，你不觉得很奇怪吗？
如果有了 onoff，你只要定义 _registerEvent 成
```
    _registerEvent: function (register) {
        this.node.onoff(register, cc.Node.EventType.TOUCH_START, this._onTouchBegan, this);
        this.node.onoff(register, cc.Node.EventType.TOUCH_MOVE, this._onTouchMove, this);
        this.node.onoff(register, cc.Node.EventType.TOUCH_END, this._onTouchEnded, this);
        this.node.onoff(register, cc.Node.EventType.TOUCH_CANCEL, this._onTouchCancel, this);

        this.node.onoff(register, cc.Node.EventType.MOUSE_ENTER, this._onMouseMoveIn, this);
        this.node.onoff(register, cc.Node.EventType.MOUSE_LEAVE, this._onMouseMoveOut, this);
    },
```

zilongshanren replied:
@jareguo  一点也不奇怪啊， register 和 unregister 一看函数名字就知道是干啥的。
重复代码不算呀，里面调用的 api 不一样。 关键是你多了一个 onoff，以后大家可能就都会用这个 api 了，添加 api 需要谨慎。另外，这种 onoff 的封装，其他开发者可以自己做，没必要引擎提供。

至于包体积，这里差不了 0.001kb 吧？

jareguo replied:
问题是你没 unregister 这个方法啊，你没提供 onoff 的话，大家就不会下意识的写出可靠、简洁的代码。否则你 CCButton 为什么不写成 unregister ？这个 API 可能大家都会用到，那用就用啊，引擎不就是鼓励大家尽量少犯错吗？

zilongshanren replied:
@jareguo Button那边是需要封装一个 unregister 的，之前写的不是很好。
我主要是觉得这个函数的参数数量有点吓人。

jareguo replied:
你会这么纠结函数参数数量的人，都能不小心“写得不是很好”，换做其它人也难免一样。那不就是说明了 onoff 的必要吗…… 如果有 onoff，大家一开始就能写得很好看。
至于参数数量，这个应该也还好吧…… 一般只用写前 4 个

pandamicro replied:
我同意子龙的观点，这个 API 太怪了

1. 没有提供任何新的功能
2. 指意不明，一眼看上去都不知道是在 on 还是在 off，代码阅读起来也有障碍
3. 如果要节省包体，用户完全可以自己封装一个类似的函数，引擎提供会给用户更多困扰，什么时候该用 onoff，什么时候该用 on & off？
4. 我也不认为提供了 onoff 对用户的代码有什么指导作用，毕竟用户需要写成 registerEvent 这种形式才能节省代码，然而用户很多时候写的代码都是 on / off 直接混在各种 onEnable / onDisable 代码中，甚至不想干的外部代码中。

其实从 API 形式上，我认为任何通过 flag 来区分功能的 API 都很可能是坏的设计，它把功能隐含在参数中，导致表意不清，功能混杂

zilongshanren commented on cocos2d/core/CCNode.js:
```diff
@@ -1204,6 +1246,10 @@ var Node = cc.Class({
         }
+
+        if (type === WORLD_POSITION_CHANGED) {
+            this._registerWPosChanged(false);
```
直接使用 register 和 unregister 不好一些吗？ @jareguo  为什么一定要通过参数来实现一个函数的多种行为呢。。

zilongshanren replied:
@jareguo  @knoxHuang  我比较倾向于一个函数只做一件事情。

jareguo replied:
代码少了，包体就小了，出错的几率就小了（严格保证 on 和 off 互相配对）

jareguo replied:
比如上面说的 CCButton 的问题，你怎么保证 onDisable 时 off 的事件，完全和 registerEvent 中的匹配？换一个人来维护，可能不小心就弄错了

zilongshanren replied:
@jareguo 并不能严格保证 on 和 off 互相配对吧，还是要通过传 true 和 false 来保证配对

jareguo replied:
这不就够了吗？你换成自动 off 的机制，只会带来复杂度或者内存占用的上升

zilongshanren replied:
@jareguo 我不是说要自动 off，我是说手动去调用 off，就像现在你要调用 onoff(false)一样。

jareguo replied:
不一样啊，你原来代码写两处，现在写一处，就能保证 off 的数量和参数一致。比如 CCButton 那边，6 个事件，每个事件参数都不一样，你写两边不蛋疼吗…… 如果另一个哪天改了 on 或 off，就有可能忘了同步另一边的 off 或 on，毕竟你是分开写的。

zilongshanren replied:
@jareguo 参考 H5 的 DOM 的 api 吧，好像没有 onoff?

pandamicro replied:
> 比如上面说的 CCButton 的问题，你怎么保证 onDisable 时 off 的事件，完全和 registerEvent 中的匹配？换一个人来维护，可能不小心就弄错了

我觉得咱们不能这么惯开发者，更不能这么惯自己的研发，维护过程中可能产生问题的地方多了去了，重要的是相信开发者的自身素养，当然，也要相信，能力有限的开发者，你再保护他也会犯一样多的错误。比如他仍然可能在不同的地方对不同的事件调用 onoff 导致不匹配而不是使用你所展示的封装。

另外，反对这种理念的另一个原因是，如果开发者自己把事件匹配错了，他不会怪引擎的，明显是自己的原因，所以这种帮衬是一种越俎代庖式的多余

zilongshanren commented on cocos2d/core/CCNode.js:
```diff
@@ -1091,6 +1093,42 @@ var Node = cc.Class({
+    _registerWPosChangedToNewParent (event) {
+        // remove oldParent event
+        if (event.detail) {
+            this._doRegisterWPosChangedTo(event.detail, false);
```
_doRegisterWPosChangedTo 和 _onWPosChanged封装感觉把代码逻辑变复杂了。。。  因为函数名字不好看，不能一眼看出这个函数的意图。不如直接 on/off 来得清晰。

jareguo replied:
_onWPosChanged 里面就一行代码，还能怎么改……
_doRegisterWPosChangedTo 你希望怎么改？

zilongshanren replied:
@jareguo  _onWPosChanged 不要啊，直接 emit 就可以了，emit 的意图很清晰了。
_doRegisterWPosChangedTo 这个名字你说是啥意思？还要结合函数参数。。。而且必须读里面的代码才知道这个函数是干啥的。我建议改成两个函数，一个是 registerWorldPosChanged，一个是 unregisterWorldPosChanged
WPos 这是啥简写啊。。。

jareguo replied:
> _onWPosChanged 不要啊，直接 emit 就可以了，emit 的意图很清晰了。

这是回调函数啊，怎么直接 emit

> WPos 这是啥简写啊。。。

这是为了统一缩写，因为有 _registerWPosChangedToNewParent 这个方法，所以 WorldPosition 简写成 WPos，这里为了保持统一。

pandamicro reviewed (changes requested):
当然，比起 API 的担忧，我同样担忧这个事件的用例。这个功能反映的是对 world transform 随时跟踪的需求，这种需求我觉得可以考虑从 renderer 层面提供，renderer 是最直接可以获知 world transform 修改以及获取新值的。目前的实现看上去太过于复杂了

pandamicro commented on cocos2d/core/CCNode.js:
```diff
@@ -1091,6 +1093,42 @@ var Node = cc.Class({
+    },
+
+    _doRegisterWPosChangedTo (target, register) {
+        target.onoff(register, WORLD_POSITION_CHANGED, this._onWPosChanged, this);
```
这里注册 parent 的 WORLD_POSITION_CHANGED，然后回调中在自己身上 emit WORLD_POSITION_CHANGED 的原因是什么？

另外，原始 parent 的 WORLD_POSITION_CHANGED 我没有看到被发出的地方

@knoxHuang

pandamicro replied:
看了一下大概理解了，这里应该监听的是父节点的 POSITION_CHANGED，所以在这里 emit WORLD_POSITION_CHANGED 并不是很严谨，因为 global transform 还没有计算，父节点
 TRS 的修改不意味着子节点的 world transform 一定会改变。

举个例子，开放了这个事件给用户之后，他可能会监听这个事件，然后在事件中用 node.getNodeToWorldTransform() 来获取 world transform 做计算，然而这个事件的时机导致它获取的 world transform 并不是当前帧的正确值

感觉是不是应该叫 PARENT_TRS_CHANGED？虽然这个意义也很奇怪，不过表述是准确的。

knoxHuang replied:
这个上面的 _registerWPosChanged 函数中

```js
this.onoff(register, POSITION_CHANGED, this._onWPosChanged, this);
```
当节点发生坐标变换的时候，就回去 emit WORLD_POSITION_CHANGED 消息

目前是节点遇到 position, scale,  rotation,  parent 改变时，回去 emit WORLD_POSITION_CHANGED 消息

pandamicro replied:
但是你现在的实现监听的是父节点的 WORLD_POSITION_CHANGED 事件
